### PR TITLE
[SPARK-29469][Shuffle] Avoid retries by RetryingBlockFetcher when ExternalBlockStoreClient is closed

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

When ExternalBlockStoreClient was closed, retries from RetryingBlockFetcher will cause NPE. This proposes to skip retries by RetryingBlockFetcher when ExternalBlockStoreClient is closed.

### Why are the changes needed?

When ExternalBlockStoreClient was closed, retries from RetryingBlockFetcher will cause NPE:

```
2019-10-14 20:06:16 ERROR RetryingBlockFetcher:143 - Exception while beginning fetch of 2 outstanding blocks (after 3 retries)
java.lang.NullPointerException
at org.apache.spark.network.shuffle.ExternalShuffleClient.lambda$fetchBlocks$0(ExternalShuffleClient.java:100)
at org.apache.spark.network.shuffle.RetryingBlockFetcher.fetchAllOutstanding(RetryingBlockFetcher.java:141)
at org.apache.spark.network.shuffle.RetryingBlockFetcher.lambda$initiateRetry$0(RetryingBlockFetcher.java:169)
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
```

It was happened after BlockManager and ExternalBlockStoreClient was closed due to previous errors. In this cases, RetryingBlockFetcher does not need to retry. This NPE is harmless for job execution, but is a source of misleading when looking at log. Especially for end-users.

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by viirya** [comment]
cc @cloud-fan

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java [hunk 1]
```diff
@@ -102,9 +102,14 @@ public void fetchBlocks(
...
-            TransportClient client = clientFactory.createClient(host, port);
-            new OneForOneBlockFetcher(client, appId, execId,
-              blockIds1, listener1, conf, downloadFileManager).start();
+            // Unless this client is closed.
+            if (clientFactory != null) {
+              TransportClient client = clientFactory.createClient(host, port);
+              new OneForOneBlockFetcher(client, appId, execId,
+                      blockIds1, listener1, conf, downloadFileManager).start();
+            } else {
+              logger.warn("This client was closed. Skipping further block fetch retries.");
```
Comment (dongjoon-hyun) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java]: Do we need to use `WARN` level? It looks a little high.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java [hunk 2]
```diff
@@ -102,9 +102,14 @@ public void fetchBlocks(
     try {
       RetryingBlockFetcher.BlockFetchStarter blockFetchStarter =
           (blockIds1, listener1) -> {
-            TransportClient client = clientFactory.createClient(host, port);
-            new OneForOneBlockFetcher(client, appId, execId,
-              blockIds1, listener1, conf, downloadFileManager).start();
+            // Unless this client is closed.
```
Comment (dongjoon-hyun) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java]: `this client` -> `this clientFactory`?

Comment (viirya) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java]: `This client` here means ExternalBlockStoreClient, not the clientFactory.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java (hunk 1 above)
Comment (viirya) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java]: info? debug?

Comment (dongjoon-hyun) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java]: I guess `INFO` is enough.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java (hunk 2 above)
Comment (dongjoon-hyun) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java]: Does it match with the code `if (clientFactory != null) {`?

Comment (viirya) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java]: I am fine for clientFactory, although I concern it sounds a bit not easy to understand, for end users.

Comment (viirya) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java]: Changed to clientFactory.

**Comment by viirya** [comment]
retest this please

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java [hunk 3]
```diff
@@ -102,9 +102,14 @@ public void fetchBlocks(
...
+            if (clientFactory != null) {
```
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java]: shall we mark `clientFactory` as `transient`? Seems like the thread to close the client and the thread to retry can be different.

Comment (viirya) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java]: you mean volatile?

Comment (cloud-fan) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java]: ah sorry, `volatile`

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by viirya** [comment]
thanks @dongjoon-hyun @cloud-fan!

**Pull request closed without merging**